pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,494 @@
+// for those leases, favoring candidates with higher versions.
+type Controller struct {
+	leaseInformer coordinationv1informers.LeaseInformer
+	leaseSynced   cache.InformerSynced
+	leaseClient   coordinationv1client.CoordinationV1Interface
+
+	leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer
+	leaseCandidateSynced   cache.InformerSynced
+	leaseCandidateClient   coordinationv1alpha1client.CoordinationV1alpha1Interface
+
+	electionQueue workqueue.TypedRateLimitingInterface[leaderLeaseID]
+}
+
+type leaderLeaseID struct {
+	namespace, name string
+}
+
+func (l leaderLeaseID) String() string {
+	return l.namespace + "/" + l.name
+}
+
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.electionQueue.ShutDown()
+
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.leaseSynced, c.leaseCandidateSynced) {
+		return
+	}
+	klog.Infof("Workers: %d", workers)
+	for i := 0; i < workers; i++ {
+		klog.Infof("Starting worker")
+		go wait.UntilWithContext(ctx, c.runElectionWorker, time.Second)
+	}
+	<-ctx.Done()
+}
+
+func NewController(leaseInformer coordinationv1informers.LeaseInformer, leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer, leaseClient coordinationv1client.CoordinationV1Interface, leaseCandidateClient coordinationv1alpha1client.CoordinationV1alpha1Interface) (*Controller, error) {
+	c := &Controller{
+		leaseInformer:          leaseInformer,
+		leaseCandidateInformer: leaseCandidateInformer,
+		leaseClient:            leaseClient,
+		leaseCandidateClient:   leaseCandidateClient,
+
+		electionQueue: workqueue.NewTypedRateLimitingQueueWithConfig(workqueue.DefaultTypedControllerRateLimiter[leaderLeaseID](), workqueue.TypedRateLimitingQueueConfig[leaderLeaseID]{Name: controllerName}),
+	}
+	leaseSynced, err := leaseInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.processLease(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.processLease(newObj)
+		},
+		DeleteFunc: func(oldObj interface{}) {
+			c.processLease(oldObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	leaseCandidateSynced, err := leaseCandidateInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.processCandidate(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.processCandidate(newObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	c.leaseSynced = leaseSynced.HasSynced
+	c.leaseCandidateSynced = leaseCandidateSynced.HasSynced
+	return c, nil
+}
+
+func (c *Controller) runElectionWorker(ctx context.Context) {
+	klog.Infof("runElectionWorker")
+	for c.processNextElectionItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextElectionItem(ctx context.Context) bool {
+	key, shutdown := c.electionQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	err := c.stepElection(ctx, key)
+	utilruntime.HandleError(err)
```

Comment:
don't you want to handle the error and retry a few times?

https://github.com/kubernetes/kubernetes/blob/b293ca905762f895f56b813ef19e61d4755aaf02/pkg/controller/endpoint/endpoints_controller.go#L327-L361

Reply:
Errors are caused by two types of conditions:

1. state issues: for instance the spec of the candidates have invalid strategies
2. CRUD failures 

I think only case 2 should warrant retries since case 1 will repeat the same problem with the same state. If the state changes, an informer event will notify. Refactoring this to return two variables.